client/cloudevents/agentclient.go

```diff
@@ -0,0 +1,287 @@
+package cloudevents
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/google/uuid"
+
+	"k8s.io/klog/v2"
+
+	"open-cluster-management.io/api/client/cloudevents/payload"
+	"open-cluster-management.io/api/client/cloudevents/protocol"
+	"open-cluster-management.io/api/client/cloudevents/types"
+)
+
+type CloudEventAgentClient[T ResourceObject] struct {
+	contextGetter    protocol.CloudEventsContextGetter
+	subClient        cloudevents.Client
+	pubClient        cloudevents.Client
+	lister           Lister[T]
+	codecs           map[types.GroupVersionResource]Codec[T]
+	statusHashGetter StatusHashGetter[T]
+	agentID          string
+	clusterName      string
+}
+
+func NewCloudEventAgentClient[T ResourceObject](
+	contextGetter protocol.CloudEventsContextGetter,
+	subClient, pubClient cloudevents.Client,
+	lister Lister[T],
+	statusHashGetter StatusHashGetter[T],
+	agentID, clusterName string,
+) *CloudEventAgentClient[T] {
+	return &CloudEventAgentClient[T]{
+		contextGetter:    contextGetter,
+		subClient:        subClient,
+		pubClient:        pubClient,
+		lister:           lister,
+		codecs:           make(map[types.GroupVersionResource]Codec[T]),
+		statusHashGetter: statusHashGetter,
+		agentID:          agentID,
+		clusterName:      clusterName,
+	}
+}
+
+func (c *CloudEventAgentClient[T]) WithCodec(gvr types.GroupVersionResource, codec Codec[T]) {
+	c.codecs[gvr] = codec
+}
+
+// Resync the resources spec of the current agent by sending a spec resync request to sources
+func (c *CloudEventAgentClient[T]) Resync(ctx context.Context, eventGVR types.GroupVersionResource) error {
+	eventType := types.CloudEventType{
+		GroupVersionResource: eventGVR,
+		SubResource:          types.SubResourceSpec,
+		Action:               types.ResyncRequestAction,
+	}
+
+	evt := cloudevents.NewEvent()
+	evt.SetID(uuid.New().String())
+	evt.SetType(eventType.String())
+	evt.SetTime(time.Now())
+	evt.SetSource(c.agentID)
+	evt.SetExtension(types.ExtensionClusterName, c.clusterName)
+
+	objs, err := c.lister.List(types.ListOptions{Source: types.SourceAll})
+	if err != nil {
+		return err
+	}
+
+	resources := &payload.ResourceVersionList{Versions: []payload.ResourceVersion{}}
```

Comment:
Nit: can we pre-allocate the  ResourceVersion slice with length of objects?
```golang
resources := &payload.ResourceVersionList{Versions: make([]payload.ResourceVersion, 0, len(objs))}
```

this can optimize memory usage, when there are two many objects, to avoid slice resize.

Reply:
done